content/starter-configs/tutorials/laravel.md

```diff
@@ -0,0 +1,172 @@
+  webserver:
+    # This uses PHP 8.3.x with Apache: update to match your version of PHP.
+    image: tugboatqa/php:8.3-apache
+
+    # Set this as the default service. This does a few things
+    #   1. Clones the git repository into the service container
+    #   2. Exposes port 80 to the Tugboat HTTP proxy
+    #   3. Routes requests to the preview URL to this service
+    default: true
+    # Wait until the mysql service is done building.
+    depends: database
+
+    # A set of commands to run while building this service
+    commands:
+      # Commands that set up the basic preview infrastructure
+      init:
+        # Install opcache and mod-rewrite.
+        - docker-php-ext-install opcache
+        - a2enmod headers rewrite
+        # Install node
+        - apt-get update
+        - apt-get install -yq ca-certificates curl gnupg
+        - mkdir -p /etc/apt/keyrings
+        - curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o
+          /etc/apt/keyrings/nodesource.gpg
+        - echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_18.x nodistro main" |
+          tee /etc/apt/sources.list.d/nodesource.list
+        - apt-get update
+        - apt-get install -yq nodejs
+        # Copy the tugboat .env we created before
+        - cp "${TUGBOAT_ROOT}/.tugboat/.env.tugboat" "${TUGBOAT_ROOT}/.env"
+        # Link the document root to the expected path. This example links
+        # /public to the docroot
+        - ln -snf "${TUGBOAT_ROOT}/public" "${DOCROOT}"
+        # Composer install
+        - composer install --optimize-autoloader
+        # Generate the random key if you didn't set APP_KEY before.
+        - php artisan key:generate
+        # Generate the db structure with some data. Here you have several choices:
+        # 1. If you have seeders on your project, you might just run migrate with --seed.
+        # - php artisan migrate --seed
+        # 2. You might want just the structure without any test data.
+        # - php artisan migrate
+        # 3. Or you might load a database dump from somewhere else. That's up to you.
+        - php artisan migrate --seed
+        # Compile vite resources
+        - npm install
+        - npm run build
+      update:
+        # Include the APP URL in .env
+        - echo "APP_URL=${TUGBOAT_DEFAULT_SERVICE_URL_HOST}" >> "${TUGBOAT_ROOT}/.env"
+        - composer install --optimize-autoloader
+        # Clear caches.
+        - php artisan config:cache
+        # Run any pending migrations. This will ensure your data has the last
+        # migrations applied.
+        - php artisan migrate --force
+        # Ensure storage permissions
+        - chgrp -R www-data "${TUGBOAT_ROOT}/storage"
+        - find "${TUGBOAT_ROOT}/storage" -type d -exec chmod 2775 {} \;
+        - find "${TUGBOAT_ROOT}/storage" -type f -exec chmod 0664 {} \;
+        # Compile vite templates
+        - npm install
+        - npm run build
```

Comment:
```suggestion
        - chgrp -R www-data "${TUGBOAT_ROOT}/storage"
        - find "${TUGBOAT_ROOT}/storage" -type d -exec chmod 2775 {} \;
        - find "${TUGBOAT_ROOT}/storage" -type f -exec chmod 0664 {} \;
```

I don't think these are needed in both update and build, but correct me if I'm wrong, I have no laravel experience. ☺️

Reply:
tbh I've read https://docs.tugboatqa.com/building-a-preview/preview-deep-dive/how-previews-work/ 30 times and still not 100% sure what goes where.

I'm pretty clear what goes on init, but build and update are quite fuzzy for me.
If I'm right, build would run in every case (having or not a base preview) 
Update will happen if I manually refresh a preview.  But it still uses the base preview if any (or init step if none).

Basically, what we need to do is run migrate any time where there might be a code change, as a new migration (script with sequence of changes to a previous db state) could be added. In laravel those are build incrementally (e.g. add this new table, or add this new column to existing table).

Reply:
And to be clear, there is no harm on running migrate twice, as it detects which migrations are already applied.

Reply:
You're not alone, it is confusing. When a preview is built with no base preview, it runs each of the command groups in order, `init`, `update`, and then `build`. Then the state of the services is committed back to a docker image. If a new preview is built from a Base preview, it then creates containers from those docker images with the saved state, and just runs the `build` phase. 

Based on what you've said here, I think my suggested change is correct. We only want to run `migrate` once per preview on every build.

Reply:
Then I wonder if our `update` should be just empty (`build` will run in every case), or some stuff from `init` should be on `update`. But I'm pretty sure that we want the background tasks installed on init, but that will trigger them and they require the db to be already initialized, which requires e.g. a previous composer install.

Reply:
okay, so reading through the commands a bit more carefully, I think maybe you should move the `artisan migrate --seed` into the `update` phase. The update phase is what will be run on a Base Preview whenever a *Refresh* occurs.


```
    commands:
      # Commands that set up the basic preview infrastructure
      init:
        # Install opcache and mod-rewrite.
        - docker-php-ext-install opcache
        - a2enmod headers rewrite

        # Install node
        - apt-get update
        - apt-get install -yq ca-certificates curl gnupg
        - mkdir -p /etc/apt/keyrings
        - curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o
          /etc/apt/keyrings/nodesource.gpg
        - echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_18.x nodistro main" |
          tee /etc/apt/sources.list.d/nodesource.list
        - apt-get update
        - apt-get install -yq nodejs

        # Copy the tugboat .env we created before
        - cp "${TUGBOAT_ROOT}/.tugboat/.env.tugboat" "${TUGBOAT_ROOT}/.env"

        # Link the document root to the expected path. This example links
        # /public to the docroot
        - ln -snf "${TUGBOAT_ROOT}/public" "${DOCROOT}"

        # Ensure storage permissions
        - chgrp -R www-data "${TUGBOAT_ROOT}/storage"
        - find "${TUGBOAT_ROOT}/storage" -type d -exec chmod 2775 {} \;
        - find "${TUGBOAT_ROOT}/storage" -type f -exec chmod 0664 {} \;

        # Install the workers
        - mkdir -p /etc/service/webserver
        - cp .tugboat/etc/service/webserver/run /etc/service/webserver/run
        - chmod +x /etc/service/webserver/run
        
      update:
        # Composer install
        - composer install --optimize-autoloader
        # Generate the random key if you didn't set APP_KEY before.
        - php artisan key:generate
        # Generate the db structure with some data. Here you have several choices:
        # 1. If you have seeders on your project, you might just run migrate with --seed.
        # - php artisan migrate --seed
        # 2. You might want just the structure without any test data.
        # - php artisan migrate
        # 3. Or you might load a database dump from somewhere else. That's up to you.
        - php artisan migrate --seed

      build:
        - composer install --optimize-autoloader
        # Clear caches.
        - php artisan config:cache
        # Run any pending migrations. This will ensure your data has the last
        # migrations applied.
        - php artisan migrate --force
        # Compile vite templates
        - npm install
        - npm run build
```

Is there an easy demo laravel app we could create? I would love to have a demo repo for this at https://github.com/TugboatDemo. Then we can hone the above commands with a "real" project.

Reply:
> I agree with you, but if I understood etc/service/webserver/run will be running before update happens, and that would fail because it requires the db to exist already.

It will fail, but it should still allow the build to complete, at which point it will stop all containers, and then commit them to images, and recreate the containers and start them.

Reply:
Not that you needed it, but I confirm it worked.